core/integration/engine_test.go

```diff
@@ -81,6 +105,18 @@ func engineWithConfig(ctx context.Context, t *testctx.T, cfgFns ...func(context.
 	}
 }
 
+// engineClientContainerWithVersion returns a container with specific version of dagger cli
+// and connected to the given devEngine service
+func engineClientContainerWithVersion(ctx context.Context, t *testctx.T, c *dagger.Client, devEngine *dagger.Service, version string) (*dagger.Container, error) {
+	endpoint, err := devEngine.Endpoint(ctx, dagger.ServiceEndpointOpts{Port: 1234, Scheme: "tcp"})
+	if err != nil {
+		return nil, err
+	}
+	return c.Container().From(fmt.Sprintf("ghcr.io/dagger/engine:%s", version)).
```

Comment:
version >= v0.12.5 have dagger binary as part of the container image. If we want this test to support comparing version < v0.12.5, we may have to add a function to pull that from GitHub release artifacts.